src/common/system_executor.cpp

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright © 2022 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored By: Christopher James Halse Rogers <christopher.halse.rogers@canonical.com>
+ */
+
+#include "mir/system_executor.h"
+
+#include "mir/thread_name.h"
+
+#include <thread>
+#include <array>
+#include <memory>
+#include <atomic>
+#include <list>
+#include <future>
+
+#include <version>
+
+#ifdef __cpp_lib_semaphore
+#include <semaphore>
+#else
+// Hello, there! This version of libstdc++ doesn't support the C++20 feature std::binary_semaphore
+// We shall, instead, open-code the simplest possible implementation of what we use here.
+namespace std
+{
+class binary_semaphore
+{
+public:
+    binary_semaphore(int initial)
+        : raised{initial == 1}
+    {
+    }
+
+    void release()
+    {
+        std::lock_guard<decltype(mutex)> lock{mutex};
+        raised = true;
+        cv.notify_all();
+    }
+
+    void acquire()
+    {
+        std::unique_lock<decltype(mutex)> lock{mutex};
+        if (raised)
+        {
+            raised = false;
+            return;
+        }
+        cv.wait(lock, [this]() { return raised; });
+        raised = false;
+    }
+
+    bool try_acquire()
+    {
+        std::lock_guard<decltype(mutex)> lock{mutex};
+        if (raised)
+        {
+            raised = false;
+            return true;
+        }
+        return false;
+    }
+private:
+    std::mutex mutex;
+    std::condition_variable cv;
+    bool raised;
+};
+}
+#endif
+
+namespace
+{
+constexpr int const min_threadpool_threads = 4;
+
+void just_rethrow_exception()
+{
+    std::rethrow_exception(std::current_exception());
+}
+
+/* We use an atomic void(*)() rather than a std::function to avoid needing to take a mutex
+ * in exception context, as taking a mutex can itself throw an exception!
+ */
+std::atomic<void(*)()> exception_handler = &just_rethrow_exception;
```

Comment:
I don't see why we give `just_rethrow_exception()` a name. Surely this is better to inline?
```suggestion
/* We use an atomic void(*)() rather than a std::function to avoid needing to take a mutex
 * in exception context, as taking a mutex can itself throw an exception!
 */
std::atomic<void(*)()> exception_handler{[] { std::rethrow_exception(std::current_exception()); }};
```

Reply:
Aha! I was having trouble getting the lambda to decay into something the compiler would accept. This is indeed better.